File: backtesting_engine/risk_manager.py
"""
Risk management and daily limits
"""

from typing import Optional


class RiskManager:
    """Implements risk controls and daily limits"""
    
    def __init__(self, daily_max_loss: float):
        self.daily_max_loss = abs(daily_max_loss)  # Ensure positive value
        self.daily_pnl = 0.0
    
    def check_daily_limit(self, current_pnl: float) -> bool:
        """Check if daily limit is breached"""
        return current_pnl <= -self.daily_max_loss
    
    def should_close_all_positions(self, total_pnl: float) -> bool:
        """Check if all positions should be closed due to daily limit"""
        return self.check_daily_limit(total_pnl)
    
    def update_daily_pnl(self, pnl: float):
        """Update daily P&L tracking"""
        self.daily_pnl = pnl
    
    def reset_daily_tracking(self):
        """Reset daily tracking for new trading day"""
        self.daily_pnl = 0.0
    
    def get_remaining_risk_capacity(self) -> float:
        """Get remaining risk capacity for the day"""
        return max(0, self.daily_max_loss + self.daily_pnl)

File: java-backtesting-engine/src/main/java/com/backtesting/engine/RiskManager.java
package com.backtesting.engine;

/**
 * Implements risk controls and daily limits
 */
public class RiskManager {
    private final double dailyMaxLoss;
    private double dailyPnl = 0.0;

    public RiskManager(double dailyMaxLoss) {
        this.dailyMaxLoss = Math.abs(dailyMaxLoss); // Ensure positive value
    }

    /**
     * Check if daily limit is breached
     */
    public boolean checkDailyLimit(double currentPnl) {
        return currentPnl <= -dailyMaxLoss;
    }

    /**
     * Check if all positions should be closed due to daily limit
     */
    public boolean shouldCloseAllPositions(double totalPnl) {
        return checkDailyLimit(totalPnl);
    }

    /**
     * Update daily P&L tracking
     */
    public void updateDailyPnl(double pnl) {
        this.dailyPnl = pnl;
    }

    /**
     * Reset daily tracking for new trading day
     */
    public void resetDailyTracking() {
        this.dailyPnl = 0.0;
    }

    /**
     * Get remaining risk capacity for the day
     */
    public double getRemainingRiskCapacity() {
        return Math.max(0, dailyMaxLoss + dailyPnl);
    }

    // Getters
    public double getDailyMaxLoss() { return dailyMaxLoss; }
    public double getDailyPnl() { return dailyPnl; }
}